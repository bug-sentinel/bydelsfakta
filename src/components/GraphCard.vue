<template>
  <div class="card-container" :class="{ large: settings.size === 'large' }">
    <div class="card">
      <div class="tabs" role="tablist">
        <button
          role="tab"
          :aria-selected="{ true: active === index }"
          :aria-label="tab.label"
          v-for="(tab, index) in settings.tabs"
          :key="index"
          @click="activeTab(index)"
          :class="{ active: active === index }"
          class="tabs__button"
          v-text="tab.label"
        ></button>
        <div
          @keydown.escape="closeMenu()"
          v-click-outside="closeMenu"
          class="context-menu"
          role="menu"
        >
          <button
            class="context-menu__button"
            @click="showDropdown = !showDropdown"
            aria-haspopup="true"
            :aria-label="showDropdown ? 'Lukk meny' : 'Ã¥pne meny'"
          >
            <i class="material-icons">{{showDropdown ? 'close' : 'menu'}}</i>
          </button>
          <div v-if="showDropdown" class="context-menu__dropdown">
            <button
              @click="savePng(settings.tabs[active].id)"
              @keyup.enter="saveSvg(settings.tabs[active].id)"
              class="context-menu__dropdown-item"
              tabindex="0"
              aria-label="lagre graf som png"
            >
              <i class="material-icons context-menu__dropdown-item-icon">photo_size_select_actual</i>
              <span>Last ned som PNG</span>
            </button>
            <button
              class="context-menu__dropdown-item"
              aria-label="lagre graf som svg"
              tabindex="0"
              @click="saveSvg(settings.tabs[active].id)"
              @keyup.enter="saveSvg(settings.tabs[active].id)"
            >
              <i class="material-icons context-menu__dropdown-item-icon">photo_size_select_actual</i>
              <span>Lagre som SVG</span>
            </button>
          </div>
        </div>
      </div>
      <graph
        v-if="settings.tabs[active] !== undefined"
        :settings="settings.tabs[active]"
        ref="graph"
      />
    </div>
  </div>
</template>

<script>
import { saveSvgAsPng } from 'save-svg-as-png';
import Graph from './Graph.vue';
import downloadSvg from '../util/downloadSvg';

export default {
  name: 'GraphCard',
  components: { Graph },
  data() {
    return {
      active: 0,
      showDropdown: false,
    };
  },
  props: {
    settings: {
      type: Object,
      required: true,
    },
  },
  methods: {
    closeMenu() {
      if (this.showDropdown) {
        this.showDropdown = false;
      }
    },

    activeTab(index) {
      this.active = index;
    },

    saveSvg(id) {
      const filename = `${this.$route.params.bydel}_${id}.svg`;
      const svgData = this.$refs.graph.$refs.svg.outerHTML;
      downloadSvg(svgData, filename);
    },

    savePng(id) {
      const file = this.$refs.graph.$refs.svg;
      const filename = `${this.$route.params.bydel}_${id}.png`;
      saveSvgAsPng(file, filename, {
        backgroundColor: 'white',
        encoderOptions: 1,
        width: file.width.baseVal.value + 40,
        height: file.height.baseVal.value + 40,
        top: -20,
        left: -20,
      });
    },
  },
};
</script>

<style scoped lang="scss">
@import './../styles/colors';

.card-container {
  padding: 0.5em;
  flex: 50% 1 10;
  min-width: 300px;
  overflow: hidden;

  @media screen and (min-width: 950px) {
    min-width: 400px;
    max-width: 50%;
  }

  &.large {
    max-width: 100%;
    flex-basis: 100%;
  }
}

.card {
  background: white;
  border: 1px solid $color-grey-100;
  border-radius: 3px;
  box-shadow: 0 1.5px 3px rgba($color-grey-600, 0.95);
  width: 100%;
}

/* Style the tabs */
.tabs {
  background-color: $color-grey-50;
  border-bottom: 1px solid $color-grey-100;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  display: flex;
  // justify-content: space-between;
  width: 100%;

  div:first-of-type a:first-child {
    border-top-left-radius: 3px;
  }

  &__button {
    background-color: $color-grey-50;
    color: rgb(41, 40, 88);
    cursor: pointer;
    float: left;
    font-weight: bold;
    transition: background-color 0.2s;
    padding: 14px 24px;

    &.active {
      background-color: #fff;
      box-shadow: 0 0 0 1px #e0e0e0;
      cursor: default;
      position: relative;

      &:after {
        background-color: white;
        bottom: -2px;
        content: '';
        height: 2px;
        left: 0;
        position: absolute;
        right: 0;
      }
    }

    /* Change background color of tabs on hover */
    &:hover:not(.active) {
      background-color: #efefef;
      box-shadow: 0 0 0 1px #e0e0e0;
      position: relative;
    }
  }
}

.context-menu {
  position: relative;
  margin-left: auto;

  &__button {
    background-color: $color-light-blue-2;
    border-top-right-radius: 3px;
    height: 3.5rem;
    width: 4rem;
  }

  &__dropdown {
    background-color: rgb(178, 210, 216);
    position: absolute;
    right: 0;
    width: 200px;
    z-index: 1;

    &-item {
      box-shadow: 0 2px 2px 0 $color-grey-200;
      color: $color-purple;
      display: flex;
      flex-direction: row;
      font-weight: bold;
      padding: 0.5rem;
      width: 100%;

      &:hover {
        background-color: lighten($color-light-blue-2, 5%);
      }

      &-icon {
        color: rgb(41, 40, 88);
        font-size: 16px;
      }

      span {
        margin-left: 0.5rem;
      }
    }
  }
}
</style>
