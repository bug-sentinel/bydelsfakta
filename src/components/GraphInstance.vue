<template>
  <div class="graph__shadow" role="tabpanel" :class="shadowClass">
    <div
      class="graph__container"
      :class="{ hidden: mode === 'table' }"
      ref="graphContainer"
      aria-hidden="true"
      @scroll="drawShadows"
      v-dragscroll
    >
      <div class="spinner" v-if="loading">
        <img src="../assets/spinner.svg" alt="" />
        <span class="spinner__text">{{ $t('loading.loadingData') }}</span>
      </div>
      <div class="error" v-if="error">
        <h2>{{ errorMessage }}</h2>
      </div>
      <svg
        v-if="!error"
        class="graph__svg"
        aria-hidden="true"
        ref="svg"
        :class="{ loading }"
        @click="showHelp = false"
      ></svg>
    </div>
    <div
      :class="{ 'visually-hidden': mode === 'graph' }"
      class="graph__tablecontainer"
      ref="tableContainer"
      v-dragscroll
      @scroll="drawShadows"
    >
      <table>
        <caption></caption>
        <thead></thead>
        <tbody></tbody>
      </table>
    </div>
    <!-- <table class="visually-hidden"> -->
    <resize-observer @notify="handleResize"></resize-observer>
    <div class="help" v-if="settings.help && mode === 'graph'">
      <button class="help__button" @click="showHelp = !showHelp">
        <i aria-hidden="true" class="material-icons">help_outline</i>
      </button>
      <div v-if="showHelp" class="help__text">{{ settings.help }}</div>
    </div>
  </div>
</template>

<script>
import * as d3 from 'd3';
import { mapState } from 'vuex';
import { dragscroll } from 'vue-dragscroll';
import TemplateA from '../util/graph-templates/templateA';
import TemplateB from '../util/graph-templates/templateB';
import TemplateC from '../util/graph-templates/templateC';
import TemplateD from '../util/graph-templates/templateD';
import TemplateE from '../util/graph-templates/templateE';
import TemplateF from '../util/graph-templates/templateF';
import TemplateG from '../util/graph-templates/templateG';
import TemplateH from '../util/graph-templates/templateH';
import TemplateI from '../util/graph-templates/templateI';
import TemplateJ from '../util/graph-templates/templateJ';
import districtNames from '../config/districtNames';

export default {
  data: () => ({
    svg: false,
    shadow: {
      left: false,
      right: false,
    },
    data: null,
    loading: true,
    currentTemplate: false,
    allDistricts: false,
    error: false,
    errorMessage: '',
    showHelp: false,
  }),

  computed: {
    shadowClass() {
      let str = '';
      if (this.shadow.left && this.mode === 'graph') str += ' graph__shadow--left';
      if (this.shadow.right) str += ' graph__shadow--right';
      return str;
    },
    ...mapState(['districts', 'compareDistricts']),
    filteredData() {
      if (!this.compareDistricts || this.districts.includes('alle')) return this.data;

      const selectedDistrictNames = this.districts.map(id => districtNames[id]);

      return {
        meta: this.data.meta,
        data: this.data.data.filter(d => {
          return selectedDistrictNames.includes(d.geography) || d.totalRow;
        }),
      };
    },
  },

  watch: {
    settings: function() {
      this.draw();
    },
    districts: function(to, from) {
      if (from.length > 1 && (to.length > 1 || this.compareDistricts)) {
        this.draw({ keepData: true });
      } else {
        this.draw();
      }
    },
  },

  mounted() {
    this.draw();
  },

  methods: {
    drawShadows() {
      let ref;
      if (this.mode === 'graph') {
        ref = 'graphContainer';
      } else if (this.mode === 'table') {
        ref = 'tableContainer';
      } else {
        return;
      }

      const width = this.$refs[ref].parentElement.clientWidth;
      const spaceLeft = this.$refs[ref].scrollLeft;
      const spaceRight = this.$refs[ref].scrollWidth - width - spaceLeft;

      this.shadow.left = spaceLeft > 5;
      this.shadow.right = spaceRight > 5;
    },

    handleResize() {
      if (this.loading) return;
      this.svg.resize(this.data, { method: this.settings.method });
      this.drawShadows();
    },

    async draw(options = {}) {
      this.loading = true;

      const geoParam = this.compareDistricts ? '00' : this.districts[0];
      if (!options.keepData) {
        this.data = await d3
          .json(`${this.settings.url}?geography=${geoParam}`)
          .then(data => {
            data.data.map(district => {
              district.geography = districtNames[district.geography];
              return district;
            });
            return data;
          })
          .catch(() => {
            this.error = true;
            this.errorMessage = this.$t('error.connectionLost');
            this.loading = false;
          });
      }

      if (this.data) {
        this.error = false;
        this.errorMessage = '';
      } else {
        return;
      }

      if (this.currentTemplate !== this.settings.template && !options.keepData) {
        switch (this.settings.template) {
          case 'a':
            this.svg = new TemplateA(this.$refs['svg']);
            break;
          case 'b':
            this.svg = new TemplateB(this.$refs['svg']);
            break;
          case 'c':
            this.svg = new TemplateC(this.$refs['svg']);
            break;
          case 'd':
            this.svg = new TemplateD(this.$refs['svg']);
            break;
          case 'e':
            this.svg = new TemplateE(this.$refs['svg']);
            break;
          case 'f':
            this.svg = new TemplateF(this.$refs['svg']);
            break;
          case 'g':
            this.svg = new TemplateG(this.$refs['svg']);
            break;
          case 'h':
            this.svg = new TemplateH(this.$refs['svg']);
            break;
          case 'i':
            this.svg = new TemplateI(this.$refs['svg']);
            break;
          case 'j':
            this.svg = new TemplateJ(this.$refs['svg']);
            break;
          default:
            break;
        }
      }
      this.currentTemplate = this.settings.template;

      this.loading = false;

      this.svg.render(this.filteredData, {
        method: this.settings.method,
        initialRender: true,
        compareDistricts: this.compareDistricts,
      });
      this.drawShadows();
    },
  },
  props: {
    settings: {
      type: Object,
      required: true,
    },
    mode: {
      type: String,
      required: true,
    },
  },
  directives: {
    dragscroll,
  },
};
</script>

<style lang="scss" scoped>
@import './../styles/colors';
@import './../styles/variables';

.graph {
  &__shadow {
    position: relative;

    &::before,
    &::after {
      bottom: 0;
      content: '';
      display: block;
      opacity: 0;
      pointer-events: none;
      position: absolute;
      top: 0;
      transition: opacity 0.1s ease-in-out;
      width: 2em;
      z-index: 1;
    }

    &::before {
      background: linear-gradient(to right, rgba(black, 0.3), rgba(black, 0.1) 60%, rgba(black, 0) 100%);
      left: 0;
    }

    &::after {
      background: linear-gradient(to left, rgba(black, 0.3), rgba(black, 0.1) 60%, rgba(black, 0) 100%);
      right: 0;
    }

    &--left::before {
      opacity: 1;
    }

    &--right::after {
      opacity: 1;
    }
  }

  &__container {
    overflow: auto;
    padding: 1em 1em 0 1em;
    position: relative;
  }

  &__tablecontainer {
    overflow-x: auto;
    position: relative;
    width: 100%;
  }

  &__svg {
    &.loading {
      animation-delay: 0.5s;
      animation-duration: 1s;
      animation-fill-mode: forwards;
      animation-iteration-count: 1;
      animation-name: fadeOut;
      pointer-events: none;
    }
  }
}

.spinner {
  align-items: center;
  animation-delay: 0.5s;
  animation-duration: 1s;
  animation-fill-mode: forwards;
  animation-iteration-count: 1;
  animation-name: fadeIn;
  background-color: rgba(black, 0.05);
  display: flex;
  flex-direction: column;
  height: 100%;
  justify-content: center;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 999;

  &__text {
    padding: 0.75em;
  }
}

.tick text {
  color: $color-grey-600;
  font-family: monospace;
  font-size: $font-small;
  font-weight: 500;
  text-rendering: geometricPrecision;
}

.help {
  position: absolute;
  top: 0.5em;
  right: 0.5em;
  z-index: 1;

  &__button {
    right: 0;
    top: 0;
    height: 3em;
    width: 3em;
    display: flex;
    border-radius: 50%;
    position: absolute;
    color: $color-purple;
    cursor: pointer;
    align-items: center;
    justify-content: center;

    & > i {
      pointer-events: none;
      font-size: 1.5em;
    }

    &:hover {
      background: $color-blue;
    }

    &::before {
      content: '';
      display: block;
      position: absolute;
      top: -0.5rem;
      right: -0.5rem;
      height: 4rem;
      width: 4rem;
      background: white;
      z-index: -1;
    }
  }

  &__text {
    padding: 1em;
    background: $color-purple;
    color: $color-yellow;
    font-weight: 500;
    position: absolute;
    top: 3em;
    right: 0;
    border-radius: 3px;
    width: 250px;
    box-shadow: 0 3px 4px rgba($color-purple, 0.5);

    &::before {
      border-bottom: 0.75em solid $color-purple;
      border-left: 0.75em solid transparent;
      border-right: 0.75em solid transparent;
      top: -0.5em;
      content: '';
      display: block;
      position: absolute;
      right: 0.75em;
    }
  }
}

@keyframes fadeIn {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

@keyframes fadeOut {
  from {
    opacity: 1;
  }

  to {
    opacity: 0;
  }
}
</style>
