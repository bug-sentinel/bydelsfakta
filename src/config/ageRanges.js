export default [
  { range: '[0, 119]', label: 'Alle', selected: true },
  { range: '[0, 1]', label: '0–1 år' },
  { range: '[0, 5]', label: '0–5 år' },
  { range: '[6, 12]', label: '6–12 år' },
  { range: '[6, 15]', label: '6–15 år' },
  { range: '[6, 17]', label: '6–17 år' },
  { range: '[12, 15]', label: '12–15 år' },
  { range: '[16, 18]', label: '16–18 år' },
  { range: '[19, 29]', label: '19–29 år' },
  { range: '[30, 49]', label: '30–49 år' },
  { range: '[50, 66]', label: '50–66 år' },
  { range: '[67, 119]', label: '67+ år' },
  { range: '[60, 79]', label: '60–79 år' },
  { range: '[80, 119]', label: '80+ år' },
  { range: '[80, 89]', label: '80–89 år' },
  { range: '[90, 119]', label: '90+ år' },
];
