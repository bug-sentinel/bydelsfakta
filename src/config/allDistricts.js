const allDistricts = [
  {
    key: '01',
    value: 'Bydel Gamle Oslo',
    uri: 'gamleoslo',
  },
  {
    key: '02',
    value: 'Bydel Grünerløkka',
    uri: 'grunerlokka',
  },
  {
    key: '03',
    value: 'Bydel Sagene',
    uri: 'sagene',
  },
  {
    key: '04',
    value: 'Bydel St. Hanshaugen',
    uri: 'sthanshaugen',
  },
  {
    key: '05',
    value: 'Bydel Frogner',
    uri: 'frogner',
  },
  {
    key: '06',
    value: 'Bydel Ullern',
    uri: 'ullern',
  },
  {
    key: '07',
    value: 'Bydel Vestre Aker',
    uri: 'vestreaker',
  },
  {
    key: '08',
    value: 'Bydel Nordre Aker',
    uri: 'nordreaker',
  },
  {
    key: '09',
    value: 'Bydel Bjerke',
    uri: 'bjerke',
  },
  {
    key: '10',
    value: 'Bydel Grorud',
    uri: 'grorud',
  },
  {
    key: '11',
    value: 'Bydel Stovner',
    uri: 'stovner',
  },
  {
    key: '12',
    value: 'Bydel Alna',
    uri: 'alna',
  },
  {
    key: '13',
    value: 'Bydel Østensjø',
    uri: 'ostensjo',
  },
  {
    key: '14',
    value: 'Bydel Nordstrand',
    uri: 'nordstrand',
  },
  {
    key: '15',
    value: 'Bydel Søndre Nordstrand',
    uri: 'sondrenordstrand',
  },
];

export default allDistricts;
