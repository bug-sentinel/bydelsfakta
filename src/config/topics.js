import alder from './topics/alder';
import fodteDodeFlytting from './topics/fodteDodeFlytting';
import boligpriser from './topics/boligpriser';
import bygningstyper from './topics/bygningstyper';
import eierform from './topics/eierform';
import befolkningsutvikling from './topics/befolkningsutvikling';
import husholdninger from './topics/husholdninger';
import innvandrerbefolkningen from './topics/innvandrerbefolkningen';
import levekaar from './topics/levekaar';
import romPerPerson from './topics/romPerPerson';

export const topicNames = [
  'befolkningsutvikling',
  'fodte-dode-flytting',
  'alder',
  'innvandrerbefolkningen',
  'husholdninger',
  'levekaar',
  'eierform',
  'bygningstyper',
  'boligpriser',
  'rom-per-person',
];

export const disabledTopics = [];

export const categories = [
  {
    kategori: 'Befolkning',
    color: 'rgb(182, 63, 50)',
    links: ['befolkningsutvikling', 'fodte-dode-flytting', 'alder', 'innvandrerbefolkningen', 'husholdninger'],
  },
  {
    kategori: 'Boforhold',
    color: 'rgb(27, 173, 120)',
    links: ['rom-per-person', 'eierform', 'bygningstyper', 'boligpriser'],
  },
  {
    kategori: 'Levek√•r',
    color: 'rgb(219, 160, 52)',
    links: ['levekaar'],
  },
];

export const topics = {
  alder,
  'fodte-dode-flytting': fodteDodeFlytting,
  boligpriser,
  bygningstyper,
  eierform,
  befolkningsutvikling,
  husholdninger,
  innvandrerbefolkningen,
  levekaar,
  'rom-per-person': romPerPerson,
};
