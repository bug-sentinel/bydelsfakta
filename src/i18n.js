import Vue from 'vue';
import VueI18n from 'vue-i18n';

function loadLocaleMessages() {
  const locales = require('./locales/nb-no.json');
  return {
    'nb-no': locales,
  };
}

export default function setupI18n() {
  Vue.use(VueI18n);

  return new VueI18n({
    locale: process.env.VUE_APP_I18N_LOCALE || 'nb-no',
    fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'nb-no',
    messages: loadLocaleMessages(),
  });
}
