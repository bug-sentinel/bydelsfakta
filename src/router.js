import Vue from 'vue';
import Router from 'vue-router';
import Home from './views/Home.vue';
import Bydel from './views/Bydel.vue';
import Alder from './views/subviews/Alder.vue';
import Boligpriser from './views/subviews/Boligpriser.vue';
import Bygningstyper from './views/subviews/Bygningstyper.vue';
import Eierform from './views/subviews/Eierform.vue';
import Flytting from './views/subviews/Flytting.vue';
import Folkemengde from './views/subviews/Folkemengde.vue';
import Husholdninger from './views/subviews/Husholdninger.vue';
import Innvandrerbefolkningen from './views/subviews/Innvandrerbefolkningen.vue';
import Kommunaleboliger from './views/subviews/Kommunaleboliger.vue';
import Levekaar from './views/subviews/Levekaar.vue';
import Trangboddhet from './views/subviews/Trangboddhet.vue';

import Graph from './components/Graph.vue';

Vue.use(Router);

export default new Router({
  routes: [
    {
      path: '/',
      name: 'Home',
      component: Home,
    },
    {
      path: '/graphA',
      name: 'Graph1',
      props: { settings: { url: '/mockDataA.json', template: 'a' } },
      component: Graph,
    },
    {
      path: '/graphB',
      name: 'Graph2',
      props: { settings: { url: '/mockDataA2.json', template: 'a' } },
      component: Graph,
    },
    {
      path: '/graphC',
      name: 'Graph3',
      props: { settings: { url: '/mockDataB.json', template: 'b' } },
      component: Graph,
    },
    {
      path: '/bydel/:bydel',
      name: 'Bydel',
      component: Bydel,
    },
    {
      path: '/bydel/:bydel/folkemengde',
      name: 'BydelFolkemengde',
      component: Folkemengde,
    },
    {
      path: '/bydel/:bydel/alder',
      name: 'BydelAlder',
      component: Alder,
    },
    {
      path: '/bydel/:bydel/innvandrerbefolkningen',
      name: 'BydelInnvandrerbefolkningen',
      component: Innvandrerbefolkningen,
    },
    {
      path: '/bydel/:bydel/husholdninger',
      name: 'BydelHusholdninger',
      component: Husholdninger,
    },
    {
      path: '/bydel/:bydel/trangboddhet',
      name: 'BydelTrandboddhet',
      component: Trangboddhet,
    },
    {
      path: '/bydel/:bydel/eierform',
      name: 'BydelEierform',
      component: Eierform,
    },
    {
      path: '/bydel/:bydel/kommunaleboliger',
      name: 'BydelKommunaleboliger',
      component: Kommunaleboliger,
    },
    {
      path: '/bydel/:bydel/flytting',
      name: 'BydelFlytting',
      component: Flytting,
    },
    {
      path: '/bydel/:bydel/bygningstyper',
      name: 'BydelBygningstyper',
      component: Bygningstyper,
    },
    {
      path: '/bydel/:bydel/boligpriser',
      name: 'BydelBoligpriser',
      component: Boligpriser,
    },
    {
      path: '/bydel/:bydel/levekaar',
      name: 'Bydellevekaar',
      component: Levekaar,
    },
    {
      path: '/sammenlign/:bydel*',
      name: 'Sammenlign',
      component: Bydel,
      children: [
        {
          path: '/folkemengde',
          name: 'Folkemengde',
          component: Folkemengde,
        },
        {
          path: '/alder',
          name: 'Alder',
          component: Alder,
        },
        {
          path: '/innvandrerbefolkningen',
          name: 'Innvandrerbefolkningen',
          component: Innvandrerbefolkningen,
        },
        {
          path: '/husholdninger',
          name: 'Husholdninger',
          component: Husholdninger,
        },
        {
          path: '/trangboddhet',
          name: 'Trandboddhet',
          component: Trangboddhet,
        },
        {
          path: '/eierform',
          name: 'Eierform',
          component: Eierform,
        },
        {
          path: '/kommunaleboliger',
          name: 'Kommunaleboliger',
          component: Kommunaleboliger,
        },
        {
          path: '/flytting',
          name: 'Flytting',
          component: Flytting,
        },
        {
          path: '/bygningstyper',
          name: 'Bygningstyper',
          component: Bygningstyper,
        },
        {
          path: '/boligpriser',
          name: 'Boligpriser',
          component: Boligpriser,
        },
        {
          path: '/levekaar',
          name: 'levekaar',
          component: Levekaar,
        },
      ],
    },
  ],
});
