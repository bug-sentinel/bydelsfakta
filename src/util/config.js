const baseUrl = () => (process.env.NODE_ENV === 'production' ? '/bydelsfakta/' : '');

export default baseUrl;
