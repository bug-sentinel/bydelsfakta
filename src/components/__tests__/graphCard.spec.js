import { mount } from '@vue/test-utils';
import { createStore } from 'vuex';
import { createRouter, createWebHistory } from 'vue-router';
import Vue3Resize from 'vue3-resize';
import VueGtag from 'vue-gtag';
import clickOutside from '@/directives/clickOutside';
import { topics } from '@/config/topics';
import mockStore from '@/../tests/MockStore';
import i18n from '@/i18n';
import { routes } from '@/router';
import GraphCard from '../GraphCard.vue';
import GraphInstance from '../GraphInstance.vue';

describe('GraphCard', () => {
  let wrapper = null;
  let router = null;
  let store = null;

  beforeEach(() => {
    router = createRouter({
      history: createWebHistory(),
      routes,
    });

    store = createStore(mockStore);

    GraphInstance.methods.draw = jest.fn();

    wrapper = mount(GraphCard, {
      global: {
        plugins: [router, store, i18n, Vue3Resize, VueGtag],
        directives: {
          'click-outside': clickOutside,
        },
        stubs: {
          'ok-icon': true,
          spinner: true,
        },
      },
      props: {
        settings: topics.alder.cards[0],
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders graphCard-component and finds card-container-class', () => {
    expect(wrapper.classes('card-container')).toBe(true);
  });

  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('change an active tab', () => {
    wrapper.setProps({ settings: topics.eierform.cards[0] });
    wrapper.vm.activeTab(1);
    expect(wrapper.vm.active).toEqual(1);
    expect(wrapper.element).toMatchSnapshot();
  });

  test('change active tab multiple times', () => {
    expect(wrapper.vm.active).toEqual(0);
    wrapper.vm.activeTab(1);
    expect(wrapper.vm.active).toEqual(1);
  });

  test('if dropdown is false, stay false', () => {
    expect(wrapper.vm.showDropdown).toEqual(false);
    wrapper.vm.closeMenu();
    expect(wrapper.vm.showDropdown).toEqual(false);
  });

  test('if dropdown is true, change to false', () => {
    wrapper.setData({ showDropdown: true });
    expect(wrapper.vm.showDropdown).toEqual(true);
    wrapper.vm.closeMenu();
    expect(wrapper.vm.showDropdown).toEqual(false);
  });

  test('saveSvg is called once', () => {
    const spy = jest.spyOn(wrapper.vm, 'saveSvg');
    spy(wrapper.vm.settings.tabs[wrapper.vm.active].id);
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
