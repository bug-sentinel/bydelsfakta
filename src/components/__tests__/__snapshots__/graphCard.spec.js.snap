// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphCard change an active tab 1`] = `
<section
  class="card-container large"
>
  <div
    class="card"
  >
    <!--v-if-->
    <header
      class="card__header"
    >
      <div
        class="card__headertext"
      >
        <h2
          class="card__title"
        >
          Befolkningen etter alder
        </h2>
      </div>
      <nav
        class="card__nav"
      >
        <div
          class="tabs"
          role="tablist"
        >
          <div
            class="vue3-resize-observer"
            tabindex="-1"
          >
            <object
              aria-hidden="true"
              data="about:blank"
              tabindex="-1"
              type="text/html"
            />
          </div>
          
          <button
            aria-label="Antall"
            class="active tabs__tab"
            data-id="tabButton-Befolkningen etter alder-0"
            role="tab"
          >
            Antall
          </button>
          <button
            aria-label="Andel"
            class="tabs__tab"
            data-id="tabButton-Befolkningen etter alder-1"
            role="tab"
          >
            Andel
          </button>
          
          <!--v-if-->
        </div>
        <div
          class="card__toggle-menu"
        >
          <button
            aria-label="Graf"
            class="card__toggle-button card__toggle-button--active"
            title="Graf"
          >
            <ok-icon-stub
              class="context-menu__dropdown-item-icon"
              iconref="graph"
              options="[object Object]"
            />
          </button>
          <button
            aria-label="Tabell"
            class="card__toggle-button"
            title="Tabell"
          >
            <ok-icon-stub
              class="context-menu__dropdown-item-icon"
              iconref="table"
              options="[object Object]"
            />
          </button>
          <!--v-if-->
        </div>
        <div
          class="context-menu"
        >
          <button
            aria-haspopup="true"
            aria-label="Åpne meny"
            class="context-menu__button"
            data-id="context-menu-button"
            title="Åpne meny"
          >
            <ok-icon-stub
              iconref="menu"
              options="[object Object]"
            />
            <span
              class="button-label"
            >
              Valg
            </span>
          </button>
          <!--v-if-->
        </div>
      </nav>
    </header>
    <div
      class="graph__shadow"
      role="tabpanel"
    >
      <div
        aria-hidden="true"
        class="graph__container"
      >
        <div
          class="spinner"
        >
          <spinner-stub />
          <span
            class="spinner__text"
          >
            Henter data
          </span>
        </div>
        <!--v-if-->
        <svg
          aria-hidden="true"
          class="graph__svg loading"
        />
      </div>
      <div
        class="visually-hidden graph__tablecontainer"
      >
        <h3
          class="table-heading"
        />
        <table
          class=""
        >
          <thead />
          <tbody />
        </table>
        <!--v-if-->
      </div>
      <div
        class="vue3-resize-observer"
        tabindex="-1"
      >
        <object
          aria-hidden="true"
          data="about:blank"
          tabindex="-1"
          type="text/html"
        />
      </div>
      <div
        class="help"
      >
        <button
          aria-label="Vis hjelpetekst"
          class="help__button"
        >
          <ok-icon-stub
            iconref="question"
            options="[object Object]"
          />
        </button>
        <!--v-if-->
      </div>
    </div>
    <!--v-if-->
    <!--v-if-->
  </div>
</section>
`;

exports[`GraphCard renders correctly 1`] = `
<section
  class="card-container large"
>
  <div
    class="card"
  >
    <!--v-if-->
    <header
      class="card__header"
    >
      <div
        class="card__headertext"
      >
        <h2
          class="card__title"
        >
          Befolkningen etter alder
        </h2>
      </div>
      <nav
        class="card__nav"
      >
        <div
          class="tabs"
          role="tablist"
        >
          <div
            class="vue3-resize-observer"
            tabindex="-1"
          >
            <object
              aria-hidden="true"
              data="about:blank"
              tabindex="-1"
              type="text/html"
            />
          </div>
          
          <button
            aria-label="Antall"
            class="active tabs__tab"
            data-id="tabButton-Befolkningen etter alder-0"
            role="tab"
          >
            Antall
          </button>
          <button
            aria-label="Andel"
            class="tabs__tab"
            data-id="tabButton-Befolkningen etter alder-1"
            role="tab"
          >
            Andel
          </button>
          
          <!--v-if-->
        </div>
        <div
          class="card__toggle-menu"
        >
          <button
            aria-label="Graf"
            class="card__toggle-button card__toggle-button--active"
            title="Graf"
          >
            <ok-icon-stub
              class="context-menu__dropdown-item-icon"
              iconref="graph"
              options="[object Object]"
            />
          </button>
          <button
            aria-label="Tabell"
            class="card__toggle-button"
            title="Tabell"
          >
            <ok-icon-stub
              class="context-menu__dropdown-item-icon"
              iconref="table"
              options="[object Object]"
            />
          </button>
          <!--v-if-->
        </div>
        <div
          class="context-menu"
        >
          <button
            aria-haspopup="true"
            aria-label="Åpne meny"
            class="context-menu__button"
            data-id="context-menu-button"
            title="Åpne meny"
          >
            <ok-icon-stub
              iconref="menu"
              options="[object Object]"
            />
            <span
              class="button-label"
            >
              Valg
            </span>
          </button>
          <!--v-if-->
        </div>
      </nav>
    </header>
    <div
      class="graph__shadow"
      role="tabpanel"
    >
      <div
        aria-hidden="true"
        class="graph__container"
      >
        <div
          class="spinner"
        >
          <spinner-stub />
          <span
            class="spinner__text"
          >
            Henter data
          </span>
        </div>
        <!--v-if-->
        <svg
          aria-hidden="true"
          class="graph__svg loading"
        />
      </div>
      <div
        class="visually-hidden graph__tablecontainer"
      >
        <h3
          class="table-heading"
        />
        <table
          class=""
        >
          <thead />
          <tbody />
        </table>
        <!--v-if-->
      </div>
      <div
        class="vue3-resize-observer"
        tabindex="-1"
      >
        <object
          aria-hidden="true"
          data="about:blank"
          tabindex="-1"
          type="text/html"
        />
      </div>
      <div
        class="help"
      >
        <button
          aria-label="Vis hjelpetekst"
          class="help__button"
        >
          <ok-icon-stub
            iconref="question"
            options="[object Object]"
          />
        </button>
        <!--v-if-->
      </div>
    </div>
    <!--v-if-->
    <!--v-if-->
  </div>
</section>
`;
